### PR TITLE
Add support for go modules

### LINKED ISSUE #74: Please support go mod
Support Go mod so that downstream projects can migrate to go mod

### PR DESCRIPTION
**Pull request by yurishkuro**
Resolves #74

- Switches to Go modules as the primary dependency manager
- Remove Glide from the build's critical paths
- Still keep Glide and Dep files for the time being, including Travis matrix steps for testing with them
- Always test with `-race`
- Use multi-package test coverage ability of `go test` (remove `cover.sh` script)

### CODE REVIEW

**Comment by yurishkuro** [comment]
cc @jpkrohling @objectiser please review

File: Makefile [hunk 1]
```diff
@@ -1,11 +1,12 @@
 PROJECT_ROOT=github.com/uber/jaeger-lib
-PACKAGES := $(shell glide novendor | grep -v ./thrift-gen/...)
+PACKAGES := $(shell go list ./... | grep -v '^github.com/uber/jaeger-lib$$' | sed 's|github.com/uber/jaeger-lib/||g' | sed 's|/.*$$||g' | sort -u | sed 's|^|./|g' | sed 's|$$|/...|g')
```
Comment (jpkrohling) [inline Makefile]: `go list ./... | awk -F\/ 'NR>1 {print "./"$4"/..."}' | sort -u` has the same effect. I find it more readable, but feel free to keep the chained sed if you think it's clearer.

Comment (yurishkuro) [reply Makefile]: nice!  Sadly, I never learned awk to the point of day-to-day usefulness.

**Comment by sagikazarmark** [comment]
This is going to be a bit problematic:

```
go get github.com/jaegertracing/jaeger-lib@v2.1.0                                                                                                                                                                           
go: finding github.com/jaegertracing/jaeger-lib v2.1.0
go: finding github.com/jaegertracing/jaeger-lib v2.1.0
go get github.com/jaegertracing/jaeger-lib@v2.1.0: github.com/jaegertracing/jaeger-lib@v2.1.0: invalid version: module contains a go.mod file, so major version must be compatible: should be v0 or v1, not v2
```

Go modules requires you to increase the major version and add it to the module path as `/v3`.

The go modules wiki contains detailed information about the process:

https://github.com/golang/go/wiki/Modules

**Comment by yurishkuro** [comment]
Thanks for the reference. I am going to revert the change for now. I think it would be best to combine the switch to go modules with changing the path from uber to jaegertracing, and start that with v1.

**Comment by sagikazarmark** [comment]
That sounds reasonable to me. IMHO you should also consider using [vanity import paths](https://sagikazarmark.hu/blog/vanity-import-paths-in-go/), like `jeagertracing.io/something`.

**Comment by jpkrohling** [comment]
Agree with everything, except with the vanity import path.

**Comment by sagikazarmark** [comment]
@jpkrohling care to elaborate? IMHO all major Go projects (or at least many of them) are moving towards vanity imports (k8s, opencensus, opentelemetry to name a few).

I've heard arguments like enterprise security making things hard for vanity imports, but proxies should make that much easier.

**Comment by jpkrohling** [comment]
I've been bitten by servers that are misconfigured or unavailable before. I'm not sure we in the Jaeger project have the bandwidth to setup/maintain this.

**Comment by sagikazarmark** [comment]
Well, proxies also help with unavailable packages. Also, import paths can be served using static site hosting (eg. Netlify). I understand the concern, but Go undoubtedly goes into this direction.

**Comment by jpkrohling** [comment]
If we have a maintainer that volunteers to set it up and run/maintain, I'm all for it :)

**Pull request merged**